tests/test_config.py

```diff
@@ -1,12 +1,80 @@
+from __future__ import annotations
 import os
 from pathlib import Path
+from typing import Any
 
 import pytest
 
 from starlette.config import Config, Environ, EnvironError
 from starlette.datastructures import URL, Secret
 
 
+def test_config_types() -> None:
+    config = Config(
+        environ={"STR": "some_str_value", "STR_CAST": "some_str_value", "BOOL": "true"}
+    )
+
+    # these functions allow us to check the return types via mypy. Using a
+    # `list` ensures that our exact types match.
+    #
+    # For example, mypy won't allow a list of type `str | None` to be passed to
+    # `is_str`. Additionally, a list of `str` can't be passed to
+    # `is_str_or_none`.
+    def is_str(x: list[str]) -> None:
+        ...
+
+    def is_bool(x: list[bool]) -> None:
+        ...
+
+    def is_int(x: list[int]) -> None:
+        ...
+
+    def is_bool_or_none(x: list[bool | None]) -> None:
+        ...
+
+    def is_str_or_none(x: list[str | None]) -> None:
+        ...
+
+    def cast_to_int(v: Any) -> int:
+        return int(v)
+
+    STR = [config("STR")]
+    is_str(STR)
+    STR_DEFAULT = [config("STR_DEFAULT", default="")]
+    is_str(STR_DEFAULT)
+    STR_CAST = [config("STR_CAST", cast=str)]
+    is_str(STR_CAST)
+    STR_NONE = [config("STR_NONE", default=None)]
+    is_str_or_none(STR_NONE)
+    STR_CAST_NONE = [config("STR_CAST_NONE", cast=str, default=None)]
+    is_str_or_none(STR_CAST_NONE)
+    STR_CAST_STR = [config("STR_CAST_STR", cast=str, default="")]
+    is_str(STR_CAST_STR)
+
+    BOOL = [config("BOOL", cast=bool)]
+    is_bool(BOOL)
+    BOOL_DEFAULT = [config("BOOL_DEFAULT", cast=bool, default=False)]
+    is_bool(BOOL_DEFAULT)
+    BOOL_NONE = [config("BOOL_NONE", cast=bool, default=None)]
+    is_bool_or_none(BOOL_NONE)
+
+    # defaults with non-bool types are allowed by our type annotations.
```

Comment:
See https://github.com/encode/starlette/discussions/1731

In a future PR, I'd like to require that the default be either `None` or the same type as the `cast`.